README.md
# Forge React Starter

[![Netlify Status](https://api.netlify.com/api/v1/badges/e0c63e91-97b5-45df-95d1-1bad86153559/deploy-status)](https://app.netlify.com/sites/forge-react-starter/deploys)

> Brings tons of thousands react libraries/components to dApps that run on [forge](https://www.arcblock.io/en/forge-sdk) powered blockchain.

[Live preview](https://forge-react-starter.netlify.com/)

![](./docs/starter-home.png)

A starter project that integrates [forge](https://docs.arcblock.io/forge/latest/) [javascript sdk](https://docs.arcblock.io/forge/latest/sdk/javascript.html) with mainstream javascript application batteries:

- [React.js](https://reactjs.org/) the framework for view layer
- [Next.js](https://nextjs.org/) for crafting pages and SSR
- [Material-UI](https://material-ui.com/) for resuable react components
- [Express.js](http://expressjs.com/) as custom web server that can serve custom api
- [Mongoose](https://mongoosejs.com/) as database ORM layer

Forge SDK libraries included in the starter project:

- [@arcblock/graphql-client](https://www.npmjs.com/package/@arcblock/graphql-client) as communication layer between application and forge powered blockchain
- [@arcblock/did-auth](https://www.npmjs.com/package/@arcblock/did-auth) help application do jwt based auth with ABT wallet
- [@arcblock/did-auth-storage-mongo](https://www.npmjs.com/package/@arcblock/did-auth-storage-mongo) storage engines that powers the magic behind ABT Wallet qrcode scanning
- [@arcblock/react-forge](https://www.npmjs.com/package/@arcblock/react-forge) react components that can implements basic UI elements to connect your application with ABT Wallet, such as avatar and auth dialog

Other javascript project goodies:

- `jwt`: token based authentication
- `eslint`: for consistent coding style
- `prettier`: for consistent code formatting
- `husky`: and `lint-staged` for prepush and precommit hooks
- `nodemon`: for auto restart server on node.js code change
- `next.js`: supports hot reload on client code change
- `dotenv`: to load configurations from `.env` files

## Folder Structure

```terminal
.
├── LICENSE
├── Makefile
├── README.md
├── app.js                    // application entry file
├── client                    // code for client side pages
│   ├── babel.config.js       // custom babel configuration
│   ├── components            // shared react components/layouts across all pages
│   ├── hooks                 // shared react hooks
│   ├── libs                  // shared utility code
│   ├── next.config.js        // custom next.js configuration
│   ├── pages                 // pages
│   └── static                // static assets that can be loaded by browser
├── package.json
├── server                    // backend code
│   ├── libs                  // shared server libs
│   ├── models                // mongoose db models
│   └── routes                // express routes and handlers
├── version
└── yarn.lock
```

## Runtime Requirements

- Mongodb v3+
- Node.js v10+
- That's all

## Usage

### Create new project with forge-cli

```terminal
npm install -g @arcblock/forge-cli forge-react-starter
forge init
forge start
forge create-project hello-forge
cd hello-forge
npm start
```

### Just use this starter repo

> **Note: You have to setup an `.env` file manually.**

```terminal
git clone https://github.com/ArcBlock/forge-react-starter.git
cd forge-react-starter
yarn
```

## Configuration

dApp configuration file is auto generated and stored in `.env`, example configure as:

```text
MONGO_URI="mongodb://127.0.0.1:27017/forge-starter"
COOKIE_SECRET="71548f479c310ebc5bd972"
CHAIN_ID="forge"
CHAIN_HOST="http://192.168.1.6:8210/api"
APP_NAME="Forge Starter"
APP_PORT="3030"
APP_SK="0xd8ee2e2b73e0f3d7b7c5d53"
APP_ID="zNKtrS7etp2WQYnbVtknbDrVa23Q3eycdcDw"
BASE_URL="http://192.168.1.6:3030"
APP_TOKEN_SECRET="5bd972379d6bdd0bd4badf49b8944994f7e2c4d3bab03c3d"
APP_TOKEN_TTL="1d"
```

> Caution: `.env` contains very sensitive info such as Application wallet secret key, PLEASE DO NOT COMMIT `.env` FILE

## FAQ

### How to upgrade `@arcblock/*` dependencies?

Simple, just update `package.json`, then `yarn`, be sure to test the starter after upgrading.

### How to deploy my application?

Checkout [Deployment.md](./docs/deployment.md)

### What APIs are supported by `GraphQLClient`?

Checkout the following screenshot or just run the starter and open browser console.

![](./docs/api-list.png)

## LICENSE

Copyright 2018-2019 ArcBlock

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## 0.9.1 (May 29, 2019)


## 0.9.0 (May 28, 2019)


## 0.8.1 (May 24, 2019)


## 0.8.0 (May 23, 2019)


## 0.7.1 (May 21, 2019)


## 0.7.0 (May 21, 2019)


## 0.6.0 (May 20, 2019)


## 0.5.1 (May 20, 2019)


## 0.5.0 (May 17, 2019)

- chore: upgrade forge sdk to v0.27.x
- chore: upgrade react-forge to v0.5.7
- chore: code refactor

## 0.4.2 (May 12, 2019)


## 0.4.0 (May 11, 2019)


## 0.3.2 (May 10, 2019)


## 0.3.1 (May 10, 2019)


## 0.3.0 (May 10, 2019)


## 0.2.3 (May 10, 2019)


## 0.2.2 (May 10, 2019)


## 0.2.1 (May 10, 2019)


## 0.2.0 (May 09, 2019)


starter.config.js
/**
 * @fileOverview Spec for forge-react-starter, can be used as a template to setup a new starter
 *
 * @requires @arcblock/forge-wallet
 */

/* eslint-disable import/order */
/* eslint-disable no-console */
/* eslint-disable object-curly-newline */
const ip = require('ip');
const fs = require('fs');
const path = require('path');
const chalk = require('chalk');
const shell = require('shelljs');
const camelCase = require('lodash/camelCase');
const { execSync } = require('child_process');
const { types } = require('@arcblock/mcrypto');
const { fromRandom, WalletType } = require('@arcblock/forge-wallet');
const { name, version } = require('./package.json');
const debug = require('debug')(name);

const defaults = {
  appName: 'Forge React Starter',
  appDescription: 'React starter application that runs on forge powered chain',
  appPort: 3030,
  mongoUri: 'mongodb://127.0.0.1:27017/forge-react-starter',
};

const questions = [
  {
    type: 'text',
    name: 'appName',
    message: 'Application name:',
    default: defaults.appName,
    validate: input => {
      if (!input) return 'Application name should not be empty';
      return true;
    },
  },
  {
    type: 'text',
    name: 'appDescription',
    message: 'Application description:',
    default: defaults.appDescription,
    validate: input => {
      if (!input) return 'Application descripiton should not be empty';
      return true;
    },
  },
  {
    type: 'text',
    name: 'appPort',
    message: 'Application listening port:',
    default: defaults.appPort,
    validate: input => {
      if (!input) return 'Application listening port should not be empty';
      return true;
    },
  },
  {
    type: 'text',
    name: 'mongoUri',
    message: 'Mongodb URI:',
    default: defaults.mongoUri,
    validate: input => {
      if (!input) return 'Mongodb connection string:';
      return true;
    },
  },
];

module.exports = {
  /**
   * Set starter version
   * @variable
   * @public
   */
  name,

  /**
   * Set starter version
   * @variable
   * @public
   */
  version,

  /**
   * Set default parameters for creating project from this starter
   * Should set defaults for all questions declared in `questions` array
   * @optional
   * @variable
   * @public
   */
  defaults,

  /**
   * List of questions used by inquire to collect parameters
   * @optional
   * @variable
   * @public
   */
  questions,

  /**
   * List of files/folders to exclude form the starter folder when creating new projects
   * @variable
   * @public
   */
  blacklist: [__filename],

  /**
   * On project folder created and files synced
   * You can create new files/modifying existing files in the project folder
   *
   * @param {object} config - project creating config
   * @param {string} config.targetDir - the project folder
   * @param {string} config.chainHost - the graphql endpoint the application running on
   * @param {string} config.chainId - the chainId of the application running on
   * @param {GraphQLClient} config.client - GraphQLClient instance that can be used to send request to the chain
   * @param {object} config.symbols - ui elements that can be used to print logs
   * @param {string} config.__starter__ - checkout `answers` for other collected parameters
   * @function
   * @public
   */
  async onConfigured(config) {
    const { chainHost, chainId, targetDir, appName, appDescription, appPort, mongoUri, client, symbols } = config;
    const ipAddress = ip.address();

    // Declare application on chain
    const wallet = fromRandom(
      WalletType({
        role: types.RoleType.ROLE_APPLICATION,
        pk: types.KeyType.ED25519,
        hash: types.HashType.SHA3,
      })
    );
    debug('application wallet', wallet.toJSON());
    const hash = await client.sendDeclareTx({
      tx: {
        chainId,
        itx: {
          moniker: camelCase(appName),
        },
      },
      wallet,
    });
    debug('application declare tx', hash);
    console.log(`${symbols.success} application account declared on chain: ${wallet.toAddress()}`);

    // Generate config
    const configPath = path.join(`${targetDir}`, '.env');
    const configContent = `MONGO_URI="${mongoUri}"
CHAIN_ID="${chainId}"
CHAIN_HOST="${chainHost.replace('127.0.0.1', ipAddress).replace('localhost', ipAddress)}"
APP_NAME="${appName}"
APP_DESCRIPTION="${appDescription}"
APP_PORT="${appPort}"
APP_SK="${wallet.secretKey}"
APP_ID="${wallet.toAddress()}"
APP_TOKEN_SECRET="${wallet.publicKey.slice(16)}"
APP_TOKEN_TTL="1d"
BASE_URL="http://${ipAddress}:${appPort}"`;
    fs.writeFileSync(configPath, configContent);
    console.log(`${symbols.success} application config generated ${configPath}`);
  },

  /**
   * On project created, files set
   * You can install dependencies for the project in this hook
   *
   * @param {object} config - project creating config
   * @param {string} config.targetDir - the project folder
   * @param {object} config.symbols - ui elements that can be used to print logs
   * @function
   * @public
   */
  onCreated(config) {
    const { targetDir, symbols } = config;
    const pm = shell.which('yarn') ? 'yarn' : 'npm';
    console.log(`${symbols.info} installing application dependencies...`);
    execSync(`cd ${targetDir} && ${pm} install`, { stdio: [0, 1, 2] });
  },

  /**
   * On project ready to run
   * You can print basic steps for users to start the application
   *
   * @param {object} config - project creating config
   * @param {string} config.targetDir - the project folder
   * @param {object} config.symbols - ui elements that can be used to print logs
   * @function
   * @public
   */
  onComplete(config) {
    const { targetDir } = config;
    const pm = shell.which('yarn') ? 'yarn' : 'npm';
    shell.echo('');
    shell.echo(chalk.cyan(`cd ${targetDir}`));
    shell.echo(chalk.cyan(`${pm} start`));
    shell.echo('');
  },
};
